api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -481,8 +534,19 @@ private NestedField(
       this.name = name;
       this.type = type;
       this.doc = doc;
-      this.initialDefault = initialDefault;
-      this.writeDefault = writeDefault;
+      this.initialDefault = castDefault(initialDefault, type);
+      this.writeDefault = castDefault(writeDefault, type);
+    }
+
+    private static Object castDefault(Object defaultValue, Type type) {
+      if (type.isNestedType() && defaultValue != null) {
+        throw new IllegalArgumentException(
+            String.format("Invalid default value for %s: %s (must be null)", type, defaultValue));
```

Comment:
Right now, we don't support non-primitive literals so this removes support for nested types. We can add that back but I think this is a good start.

Reply:
Yes. The reason not to extend this to non-primitives right now is that we will need to do even more heavy-weight work for nested types